geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java

```diff
@@ -108,11 +111,19 @@ private static void execute(ExecutablePool pool, String region, String function,
 
         isReexecute = true;
         resultCollector.clearResults();
+        failedNodes = allocateFailedNodesIfEmpty(failedNodes);
         failedNodes.clear();
       }
     } while (true);
   }
 
+  private static Set<String> allocateFailedNodesIfEmpty(Set<String> failedNodes) {
+    if (failedNodes.equals(Collections.emptySet())) {
```

Comment:
I feel like `equals()` is the wrong kind of equality for this condition.

When `failedNodes` is an `EmptySet`, this line calls `AbstractSet.equals()`. That method returns true because the two object references are equal.

When `failedNodes` is a `HashSet`, this line still calls `AbstractSet.equals()`. If the `HashSet` happens to be empty, then `equals()` returns true because the two objects have the same number of entries (0).

While it may be true that this method will never be called with an empty `HashSet`—there is nothing in the method itself to prevent that. We'd like to be able to look at this method in isolation and convince ourselves that it works.

I wonder if the condition would be clearer as:

```java
if (failedNodes == Collections.EMPTY_SET)
```

On the plus side that solves any problem of constructing a new `HashSet()` merely because the one referenced by `failedNodes` happens to be empty (and not a reference to `EMPTY_SET`. On the minus side it relies on `Collections.getEmptySet()` returning the same reference each time.

The minus isn't a really a minus in my mind since I believe pretty strongly that the whole point of the `Collections.getEmptySet()` method is to typecast the single (VM-wide) empty, immutable, set: `Collections.EMPTY_SET`.

Reply:
Swap the order of the equals check should address this.

Reply:
@Bill Good point. I agree with your proposal to solve the issue.
@pivotal-jbarrett Swapping the order of the equals in the condition would not help. It would still return true when comparing against a zero entries `HashSet`.